lms/services/application_instance.py

```diff
@@ -35,6 +36,85 @@ def __init__(self, application_instance: ApplicationInstance):
         )
 
 
+@dataclass
+class SearchBuilder:
+    model: BASE
+    filters: dict = field(default_factory=dict)
+    key_map: dict = field(default_factory=lambda: {"id_": "id", "type_": "type"})
+    query_start: Optional[callable] = None
+
+    def get_query(self, db_session, kwargs, combine=sa.and_):
+        query = self.query_start(db_session) or db_session.query(self.model)
+
+        return self.filter_query(query, kwargs, combine=combine)
+
+    def filter_query(self, query, kwargs, combine=sa.and_):
+        if clauses := list(self.get_clauses(kwargs)):
+            query = query.filter(combine(*clauses))
+
+        return query
+
+    def get_clauses(self, kwargs):
+        for key, value in kwargs.items():
+            clause = self.get_clause(key, value)
+            # Watch out as real clauses can be falsy when evaluated
+            if clause is not None:
+                yield clause
+
+    def get_clause(self, key, value):
+        if value is None:
+            return
+
+        key = self.key_map.get(key, key)
+
+        if filter_function := self.filters.get(key):
+            return filter_function(value)
+
+        if hasattr(self.model, key):
+            return getattr(self.model, key) == value
```

Comment:
This provides basic matching for attributes on the model without requiring a filter function